lib/client/src/priority_fees.rs

```diff
@@ -0,0 +1,240 @@
+        let data = self.data.read().unwrap();
+        if let Some(last_update) = data.last_update {
+            if Instant::now().duration_since(last_update) > self.config.max_age {
+                return self.config.fallback_prio;
+            }
+        } else {
+            return self.config.fallback_prio;
+        }
+        data.ema as u64
+    }
+}
+
+impl CuPercentileEmaPriorityFeeProvider {
+    pub fn run(
+        config: EmaPriorityFeeProviderConfig,
+        sender: &broadcast::Sender<BlockPrioFees>,
+    ) -> (Arc<Self>, JoinHandle<()>) {
+        let this = Arc::new(Self {
+            data: Default::default(),
+            config,
+        });
+        let handle = tokio::spawn({
+            let this_c = this.clone();
+            let rx = sender.subscribe();
+            async move { Self::run_update_job(this_c, rx).await }
+        });
+        (this, handle)
+    }
+
+    async fn run_update_job(provider: Arc<Self>, mut rx: broadcast::Receiver<BlockPrioFees>) {
+        let config = &provider.config;
+        loop {
+            let block_prios = rx.recv().await.unwrap();
+            let prio = match block_prios.by_cu_percentile.get(&config.percentile) {
+                Some(v) => *v as f64,
+                None => {
+                    error!("percentile not available: {}", config.percentile);
+                    continue;
+                }
+            };
+
+            let mut data = provider.data.write().unwrap();
+            data.ema = data.ema * (1.0 - config.alpha) + config.alpha * prio;
+            data.last_update = Some(Instant::now());
+        }
+    }
+}
+
+#[derive(Clone, Default, Debug)]
+pub struct BlockPrioFees {
+    pub slot: u64,
+    // prio fee percentile in percent -> prio fee
+    pub percentile: HashMap<u8, u64>,
+    // cu percentile in percent -> median prio fee of the group
+    pub by_cu_percentile: HashMap<u8, u64>,
+}
+
+#[derive(serde::Deserialize)]
+struct BlockPrioritizationFeesNotificationContext {
+    slot: u64,
+}
+
+#[derive(serde::Deserialize)]
+struct BlockPrioritizationFeesNotificationValue {
+    by_tx: Vec<u64>,
+    by_tx_percentiles: Vec<f64>,
+    by_cu: Vec<u64>,
+    by_cu_percentiles: Vec<f64>,
+}
+
+#[derive(serde::Deserialize)]
+struct BlockPrioritizationFeesNotificationParams {
+    context: BlockPrioritizationFeesNotificationContext,
+    value: BlockPrioritizationFeesNotificationValue,
+}
+
+fn as_block_prioritization_fees_notification(
+    notification_str: &str,
+) -> anyhow::Result<Option<BlockPrioFees>> {
+    let notification: Notification = match serde_json::from_str(&notification_str) {
+        Ok(v) => v,
+        Err(_) => return Ok(None), // not a notification at all
+    };
+    if notification.method != "blockPrioritizationFeesNotification" {
+        return Ok(None);
+    }
+    let map = match notification.params {
+        Params::Map(m) => m,
+        _ => anyhow::bail!("unexpected params, expected map"),
```

Comment:
I'm trying to understand when exactly this case would be triggered -- is it only if the lite rpc endpoint returns some respone that does not adhere to the JSON-RPC spec?

Reply:
I didn't double check the spec, but jsonrpc_core also allows array params or null params. But the notifications we expect should always have an object as params.